### PR TITLE
[VxlanOrch] pytest for EVPN VXLAN

### PR DESCRIPTION
**Pull request by srj102**
Added Pytest to test changes to VxlanMgr, VxlanOrch and PortsOrch for EVPN VXLAN feature.

HLD Location : https://github.com/Azure/SONiC/pull/437/commits

Signed-off-by: Rajesh Sankaran rajesh.sankaran@broadcom.com

What I did
Please refer to details provided above.

Why I did it
EVPN VXLAN Implementation

How I verified it
Changes as part of PR 1264 and 1266 verified with this pytest
#1264  and #1266 should be merged for this PR.

### CODE REVIEW

**Comment by srj102** [comment]
retest vs please

**Comment by srj102** [comment]
retest vs please

**Comment by srj102** [comment]
retest vs please

**Review by prsunny — APPROVED** [review]
lgtm.. request @daall to take a look

File: tests/test_evpn_tunnel.py [hunk 1]
```diff
@@ -0,0 +1,749 @@
...
+        vlanlist = ['100', '101', '102']
+        vnilist = ['1000', '1001', '1002']
+
+        print("Testing SIP Tunnel Creation")
+        vxlan_obj.check_vxlan_sip_tunnel(dvs, tunnel_name, '6.6.6.6', vlanlist, vnilist, '2.2.2.2', False)
+        print("Pass")
+
+        print("Testing Tunnel Map Entry")
+        vxlan_obj.check_vxlan_tunnel_map_entry(dvs, tunnel_name, vlanlist, vnilist)
+        print("Pass")
```
Comment (prsunny) [inline tests/test_evpn_tunnel.py:738]: Suggest removing the print("Pass") statements

File: tests/test_evpn_tunnel.py [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+
+        print("Testing SIP Tunnel Deletion")
+        remove_evpn_nvo(dvs, 'nvo1')
+        remove_vxlan_tunnel(dvs, tunnel_name)
+        vxlan_obj.check_vxlan_sip_tunnel_delete(dvs, tunnel_name)
+
+
+#    Test 3 - Create and Delete SIP Tunnel and Map entries
+#    @pytest.mark.skip(reason="Starting Vxlanmgr to be merged")
+    def test_p2mp_tunnel_with_dip(self, dvs, testlog):
```
Comment (msreddypaluru) [inline tests/test_evpn_tunnel.py:711]: what is the use of testlog, not used anywhere? can all print's can be moved to testlog?

File: tests/test_evpn_tunnel.py [hunk 3]
```diff
@@ -0,0 +1,749 @@
...
+        attrs.append(("dst_ip", dst_ip))
+
+    # create the VXLAN tunnel Term entry in Config DB
+    create_entry_tbl(
+        conf_db,
+        "VXLAN_TUNNEL", '|', name,
+        attrs,
+    )
+
+def create_vxlan_tunnel_map(dvs, tnlname, mapname, vni_id, vlan_id):
```
Comment (msreddypaluru) [inline tests/test_evpn_tunnel.py:149]: general comment - variables can be followed either map_name or mapName.

File: tests/test_evpn_tunnel.py [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+        status, fvs = tbl.get(self.tunnel_appdb[tunnel_name])
+        assert status == False, "SIP Tunnel entry not deleted"
+
+        tbl = swsscommon.Table(asic_db, self.ASIC_TUNNEL_TERM_ENTRY)
+        status, fvs = tbl.get(self.tunnelterm[tunnel_name])
+        assert status == False, "SIP Tunnel Term entry not deleted"
+
+        tbl = swsscommon.Table(asic_db, self.ASIC_TUNNEL_TABLE)
+        status, fvs = tbl.get(self.tunnel[tunnel_name])
+        assert status == False, "SIP Tunnel entry not deleted"
```
Comment (msreddypaluru) [inline tests/test_evpn_tunnel.py:306]: line 298 & 306 return same error message, it is difficult to distinguish which error is for what. Please add the db name the log.

File: tests/test_evpn_tunnel.py [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+        assert status == False, "SIP Tunnel mapper1 not deleted"
+        status, fvs = tbl.get(self.tunnel_map_map[tunnel_name][2])
+        assert status == False, "SIP Tunnel mapper2 not deleted"
+        status, fvs = tbl.get(self.tunnel_map_map[tunnel_name][3])
+        assert status == False, "SIP Tunnel mapper3 not deleted"
+
+    def check_vxlan_sip_tunnel(self, dvs, tunnel_name, src_ip, vidlist, vnidlist, dst_ip = '0.0.0.0', skip_dst_ip = 'True'):
+        asic_db = swsscommon.DBConnector(swsscommon.ASIC_DB, dvs.redis_sock, 0)
+        app_db = swsscommon.DBConnector(swsscommon.APPL_DB, dvs.redis_sock, 0)
+#        global loopback_id, def_vr_id
```
Comment (msreddypaluru) [inline tests/test_evpn_tunnel.py:321]: Can be cleaned up if not needed.

**Review by msreddypaluru — APPROVED** [review]
few minor comments, please address. approved

File: tests/test_evpn_tunnel.py (hunk 4 above)
Comment (srj102) [reply tests/test_evpn_tunnel.py]: Done

File: tests/test_evpn_tunnel.py (hunk 3 above)
Comment (srj102) [reply tests/test_evpn_tunnel.py]: Done

File: tests/test_evpn_tunnel.py (hunk 2 above)
Comment (srj102) [reply tests/test_evpn_tunnel.py:655]: test_vnet.py and test_vxlan_tunnel.py had references to testlog and hence inherited the same. 
Have removed it now

File: tests/test_evpn_tunnel.py [hunk 6]
```diff
@@ -0,0 +1,749 @@
...
+
+    tunnel_map_ids       = set()
+    tunnel_map_entry_ids = set()
+    tunnel_ids           = set()
+    tunnel_term_ids      = set()
+    bridgeport_ids       = set()
+    tunnel_map_map       = {}
+    tunnel               = {}
+    tunnel_appdb         = {}
+    tunnelterm           = {}
```
[the remark this answers is not in the document]
Comment (srj102) [reply tests/test_evpn_tunnel.py]: done

File: tests/test_evpn_tunnel.py [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+import json
+import random
+import pytest
+from pprint import pprint
+
+
+def create_entry(tbl, key, pairs):
+    fvs = swsscommon.FieldValuePairs(pairs)
+    tbl.set(key, fvs)
+    time.sleep(1)
```
[the remark this answers is not in the document]
Comment (srj102) [reply tests/test_evpn_tunnel.py:12]: This file was based on test_vnet.py and hence inherits some of those. However I have seen that sleeps are required.

File: tests/test_evpn_tunnel.py (hunk 1 above)
Comment (srj102) [reply tests/test_evpn_tunnel.py]: Done

File: tests/test_evpn_tunnel.py (hunk 2 above)
Comment (daall) [reply tests/test_evpn_tunnel.py:655]: `testlog` is a pytest fixture, it adds `test <blank> start` and `test <blank> end` to the syslogs to aid in debugging. It *should* probably just be auto-use so we don't have to explicitly invoke it everywhere, but that can be addressed in a different pr. 😄

**Comment by srj102** [comment]
@msreddypaluru @prsunny all comments are handled. Pl look through and merge.

**Review by daall — CHANGES_REQUESTED** [review]
I think the logic/flow in TestVxlanOrch makes a lot of sense and I think having all the helper methods makes it super readable, but in order to make sure the test is stable and keep the logic a little simpler we should use `dvslib` as much as possible in the helper methods.

I've pointed out some specific suggestions in my comments but in general (other than producerstatetable) you should use the methods in `DVSDatabase` instead of directly calling `swsscommon`. There are a lot of comments in `dvs_datbase.py` that should help you out and there are examples of this in action in test_acl, test_vlan, test_subport, test_route, etc.

Doing this 1) should ensure this test doesn't become flaky down the road, 2) it should improve performance because you won't have to depend on `time.sleep` to make the test behave correctly, and 3) it should make the code a lot less verbose. Let me know if you have any questions!

File: tests/test_evpn_tunnel.py [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+    ]
+
+    # create the VXLAN tunnel Term entry in Config DB
+    delete_entry_tbl(
+        conf_db,
+        "VXLAN_TUNNEL_MAP", "%s|%s" % (tnl_name, map_name),
+    )
+
+def remove_evpn_remote_vni(dvs, vlan_id, remote_vtep ):
+    app_db = swsscommon.DBConnector(swsscommon.APPL_DB, dvs.redis_sock, 0)
```
Comment (daall) [inline tests/test_evpn_tunnel.py:199]: There are also connectors you can use in `dvs` like `get_app_db` and `get_asic_db`.

**Review by daall — APPROVED** [review]
comments to be addressed in #1564

**Pull request merged**